### xgboost 

It is one of the leading algorithm that many data scientist use it because it perform better than other machine learning algorithm in two aspects 


1- Speed 

2- Performance 


why it is so ?

speed :  It follows parallism approach and cache optimization
performance- It has implicit regularization parameter and tree pruning threshold parameter which prevents it from overfitting 

So to check its two best aspects we compare the XGBoost with other Machine Learning algorithms


### Importing required libraries

In [1]:
import pandas as pd
import numpy as np
from xgboost import XGBClassifier  ## xgboost is not in sklearn library 
from sklearn.model_selection import train_test_split
#import pickle
from sklearn import datasets
import time
#from sklearn.ensemble import AdaBoostClassifier,GradientBoostingClassifier

from sklearn.metrics import accuracy_score
#from catboost import CatBoostClassifier ### to handle categorical data with out encoding them explicitely

#from lightgbm import LBMClassifier


### Import required packages of models to compare their speed and performance with xgboost

In [55]:
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC


In [3]:
#pip install xgboost

In [4]:
#pip install catboost 

In [5]:
#pip install Lightgbm

In [6]:
#pip install lightgbm

### Loading  the Data Set

In [7]:
df=pd.read_csv("D:\\python_datascience\\data sets\\diabetes.csv")
df

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
0                       0.627   50        1  
1                       0.351   31        0  
2                       0.672   32        1  
3                       0.167   21        0  
4                       2.288   33        1  
..                        ...  ...      ...  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0  

[768 rows x 9 columns]

####  Checking shape of Data

In [64]:
df.shape

(768, 9)

#### Checking missing values 

In [65]:
df.isnull().sum()

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

#### Checking class imbalance

In [66]:
df["Outcome"].value_counts()

0    500
1    268
Name: Outcome, dtype: int64

 It is showing there is no class imbalance in the data 

#### Dividing the data set into features as x and target as y

In [68]:
x=df.drop(columns=["Outcome"]) ## features which are factors of diabetes
y=df["Outcome"] ## target :   Indicates whether the person has diabetes or not 

### Dividing the dataset into training and test data as 75% as training data and 25% as test data 

In [70]:
x_train,x_test,y_train,y_test=train_test_split(x,y,random_state=3) # default : 0.25

#### Checking the shape of training data and test data 

In [11]:
print(x_train.shape)
print(x_test.shape)
print(y_train.shape)
print(y_test.shape)

(576, 8)
(192, 8)
(576,)
(192,)


#### Running Various Models

#### Comparable algorithms

1- Logistic Regression
2- KNN
3- SVM
4 -XGBoost


In [73]:

names=[]
#scoring="accuracy"

models=[]
models.append(("LogisticRegression",LogisticRegression()))
models.append(("KNN",KNeighborsClassifier()))
models.append(("SVM",SVC()))
models.append(("XGB",XGBClassifier()))
## evaluate each model in turn

for name,model in models:
        start_time=time.time()
        model.fit(x_train,y_train)
        y_pred=model.predict(x_test)  ## predictions on test data
        predictions=[round(value)for value in y_pred]
        # evaluate predictions
        accuracy=accuracy_score(y_test,predictions)  ### accuracy
        print("accuracy of ",name,":",accuracy*100)
        print("time in seconds",time.time()-start_time,"\n")
              
            

accuracy of  LogisticRegression : 72.91666666666666
time in seconds 0.03999757766723633 

accuracy of  KNN : 68.22916666666666
time in seconds 0.008000850677490234 

accuracy of  SVM : 69.79166666666666
time in seconds 0.022686004638671875 

accuracy of  XGB : 75.0
time in seconds 0.09372687339782715 



C:\Users\acer\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


### we can see XGB gives best performance out of others i.e 75% although it execution times takes more than others i.e 0.97 seconds because since the shape of the diabtetes data set is 768 rows and 9 columns so it does not give much speed on small data set it really works when we have distributed system

### Hyper parameter tuning of some parameters to get improvement in its performance 

In [76]:
xgbmodel=XGBClassifier() ## taking default parameters
xgbmodel.fit(x_train,y_train) ## model fitting

XGBClassifier(base_score=0.5, booster='gbtree', callbacks=None,
              colsample_bylevel=1, colsample_bynode=1, colsample_bytree=1,
              early_stopping_rounds=None, enable_categorical=False,
              eval_metric=None, gamma=0, gpu_id=-1, grow_policy='depthwise',
              importance_type=None, interaction_constraints='',
              learning_rate=0.300000012, max_bin=256, max_cat_to_onehot=4,
              max_delta_step=0, max_depth=6, max_leaves=0, min_child_weight=1,
              missing=nan, monotone_constraints='()', n_estimators=100,
              n_jobs=0, num_parallel_tree=1, predictor='auto', random_state=0,
              reg_alpha=0, reg_lambda=1, ...)

In [77]:
y_pred=xgbmodel.predict(x_test)  ## predictions on test data 
y_pred

array([0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1,
       0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 1, 1,
       1, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 0,
       0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0,
       0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0,
       1, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0,
       1, 1, 0, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1])

In [78]:
accuracy=accuracy_score(y_test,y_pred)  ### accuracy 
accuracy

0.75

75% accuracy

In [82]:
### hyper parameter tuning on some parameters
## learning rate  =0.1 
xgbmodel=XGBClassifier(learning_rate=0.1)
xgbmodel.fit(x_train,y_train)
y_pred=xgbmodel.predict(x_test)
accuracy=accuracy_score(y_test,y_pred)
accuracy

0.7291666666666666

In [81]:
## learning rate  =0.01 
xgbmodel=XGBClassifier(learning_rate=0.01)
xgbmodel.fit(x_train,y_train)
y_pred=xgbmodel.predict(x_test)
accuracy=accuracy_score(y_test,y_pred)
accuracy

0.71875

In [85]:
## learning_rate=0.3
xgbmodel=XGBClassifier(learning_rate=0.5) 
xgbmodel.fit(x_train,y_train)
y_pred=xgbmodel.predict(x_test)
accuracy=accuracy_score(y_test,y_pred)
accuracy

0.6979166666666666

In [86]:
xgbmodel=XGBClassifier(gamma=50) 
xgbmodel.fit(x_train,y_train)
y_pred=xgbmodel.predict(x_test)
accuracy=accuracy_score(y_test,y_pred)
accuracy

0.6666666666666666

In [87]:
xgbmodel=XGBClassifier(gamma=10) 
xgbmodel.fit(x_train,y_train)
y_pred=xgbmodel.predict(x_test)
accuracy=accuracy_score(y_test,y_pred)
accuracy

0.6822916666666666

In [93]:
xgbmodel=XGBClassifier(gamma=10,learning_rate=0.01) 
xgbmodel.fit(x_train,y_train)
y_pred=xgbmodel.predict(x_test)
accuracy=accuracy_score(y_test,y_pred)
accuracy

0.6770833333333334

By changing some parameters like learning rate and gamma we cant see any improvements in accuracy so we can do gridsearchcv to get optimal parameters on this data corresponding to xgboost